WalletWasabi/Blockchain/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -273,6 +273,33 @@ private ProcessedResult ProcessNoLock(SmartTransaction tx)
 			TransactionStore.AddOrUpdate(tx);
 		}
 
+		if (tx.Confirmed)
+		{
+			// If a key is internal and spent all its coins, then it shouldn't be used again.
+			foreach (var spenderKey in myInputs.Select(x => x.HdPubKey).Where(x => x.IsInternal).Distinct())
+			{
+				if (spenderKey.Coins.Any(x => !x.IsSpent()))
+				{
+					// The key still has unspent coins.
+					continue;
+				}
+
+				if (spenderKey.KeyState != KeyState.Spent)
+				{
+					KeyManager.SetKeyState(KeyState.Spent, spenderKey);
+					spenderKey.LatestSpentHeight = tx.Height;
+				}
+				else
+				{
+					if (spenderKey.LatestSpentHeight < tx.Height)
```

Comment:
The comment is not helpful. Instead of that, something like. 
detect if there was an address-reuse on the internal key.